### PR TITLE
Support round operation on datetime64[ns] datatypes

### LINKED ISSUE #9652: [FEA] Support `series.dt.round` and `DatetimeIndex.round` operation
**Is your feature request related to a problem? Please describe.**
Support for `ceil` was added in #8942 and `floor` in #9571 for a `Series.dt` column type. #9554 addresses adding support for `ceil` and `floor` for a `DatetimeIndex` column. Once we are able to support `ceil` and `floor` on columns of type `datetime64[ns]`, it would be great to add support for `round` operation as well. 

This can eventually be followed by issue #9561.

**Describe the solution you'd like**
`Series.dt.round`
```python
>>> rng = pandas.date_range('1/1/2018 11:59:00', periods=3, freq='min')
>>> cudf.Series(rng).dt.round("H")
0   2018-01-01 12:00:00
1   2018-01-01 12:00:00
2   2018-01-01 12:00:00
dtype: datetime64[ns]
```

`DatetimeIndex.round`
```python
>>> rng = cudf.Index(pd.date_range('1/1/2018 11:59:00', periods=3, freq='min'))
>>> rng.round('H')
DatetimeIndex(['2018-01-01 12:00:00', '2018-01-01 12:00:00',
               '2018-01-01 12:00:00'],
              dtype='datetime64[ns]', freq=None)
```

**Additional context**
https://pandas.pydata.org/docs/reference/api/pandas.Series.dt.round.html
https://pandas.pydata.org/docs/reference/api/pandas.DatetimeIndex.round.html?#pandas-datetimeindex-round

### PR DESCRIPTION
**Pull request by mayankanand007**
This PR fixes #9652, by adding support for doing `round` operation on `dtaetime64[ns]` types, which is essentially supporting `series.dt.round` and `DatetimeIndex.round`.

In addition to this, we move the round implementation that is currently there from `Frame` to `IndexedFrame` as `pd.Index` doesn't support `round`. This is why we move this implementation to `IndexedFrame`, and add the code specifically for `DatetimeIndex` (for this PR), so as to avoid having a round method for another index types.

### CODE REVIEW

**Comment by mayankanand007** [comment]
rerun tests

File: cpp/include/cudf/datetime.hpp
Comment (bdice) [inline cpp/include/cudf/datetime.hpp:473]: This doesn't necessarily round down. (applies below as well)
```suggestion
 * @brief Round to the nearest day
```

File: cpp/src/datetime/datetime_ops.cu
Comment (bdice) [inline cpp/src/datetime/datetime_ops.cu:57]: Add doxygen docstrings.
```suggestion
enum class rounding_kind {
    CEIL,  ///< Rounds up to the next integer multiple of the provided frequency
    FLOOR, ///< Rounds down to the next integer multiple of the provided frequency
    ROUND  ///< Rounds to the nearest integer multiple of the provided frequency
};
```

File: cpp/tests/datetime/datetime_ops_test.cpp
Comment (bdice) [inline cpp/tests/datetime/datetime_ops_test.cpp:978]:
```suggestion
  CUDF_TEST_EXPECT_COLUMNS_EQUAL(*round_microsecond(input), expected_microsecond);
```

File: cpp/tests/datetime/datetime_ops_test.cpp
Comment (bdice) [inline cpp/tests/datetime/datetime_ops_test.cpp:986]:
```suggestion
  CUDF_TEST_EXPECT_COLUMNS_EQUAL(*round_nanosecond(input), expected_nanosecond);
```

File: python/cudf/cudf/core/indexed_frame.py
Comment (bdice) [inline python/cudf/cudf/core/indexed_frame.py:706]: Is moving this function `round` from `Frame` to `IndexedFrame` part of the declared aim of this PR (rounding datetimes)? If so, can you explain the conflict? If not, I would do this in a separate PR.

File: python/cudf/cudf/core/indexed_frame.py
Comment (bdice) [inline python/cudf/cudf/core/indexed_frame.py:722]:
```suggestion
            or "half_up" rounding.
```

File: python/cudf/cudf/core/series.py
Comment (bdice) [inline python/cudf/cudf/core/series.py:4683]:
```suggestion
            Series with all timestamps rounded to the specified frequency.
```

File: python/cudf/cudf/core/series.py
Comment (bdice) [inline python/cudf/cudf/core/series.py:4693]: This feels like an unhelpful example. Could we construct an example that has unique results? Perhaps copy from another function's docstring.

**Review by bdice — CHANGES_REQUESTED** [review]
Here is some early-stage feedback. I would like to consider exposing the `round_general` function and `rounding_kind` enum to Cython so that we can unify our implementations at the Cython bindings level as well as the C++ level. Let's set up a call if you'd like to discuss this design proposal in more detail.

File: python/cudf/cudf/core/series.py
Comment (bdice) [reply python/cudf/cudf/core/series.py:4693]: As a very minor suggestion, I would prefer a different variable name like `dates` instead of `rng`. The word `rng` means "random number generator" to me, not "range."

File: cpp/include/cudf/datetime.hpp
Comment (mayankanand007) [reply cpp/include/cudf/datetime.hpp:473]: Thanks Bradley for the reviews! I think I get your vision: we want to introduce the same abstraction we have now in our C++ layer, in the Cython layer, sort of like a dispatcher. Am I understanding this correctly? we can certainly chat more about this offline :)

**Comment by shwina** [comment]
> Here is some early-stage feedback. I would like to consider exposing the round_general function and rounding_kind enum to Cython so that we can unify our implementations at the Cython bindings level 

I'd just caution that we don't want to expose libcudf internals to Cython. If the function is going to be part of the public libcudf API, it's fine to consume from Cython

File: cpp/src/datetime/datetime_ops.cu
Comment (jrhemstad) [inline cpp/src/datetime/datetime_ops.cu:60]: `rounding_kind::ROUND` isn't very descriptive. How about "nearest"?

File: python/cudf/cudf/core/indexed_frame.py
Comment (bdice) [reply python/cudf/cudf/core/indexed_frame.py:706]: Okay, I see why this helps. In an ideal world, the changes to `Index` would be a separate PR that gets merged before this one. If you have time, I would consider doing that. If you don't have time to split the PR, I would edit this PR description to list the reasons you gave above for why this change is being made here. PR descriptions are an excellent place to document _why_ changes are being made, because that motivation is not always clear to reviewers.

File: cpp/src/datetime/datetime_ops.cu
Comment (bdice) [reply cpp/src/datetime/datetime_ops.cu:60]: We might even want a name like `NEAREST_HALF_UP` since other parts of cuDF have support for rounding halves to the nearest even value (which this does not need to handle, to my knowledge). https://github.com/rapidsai/cudf/blob/8002cbd87367135a941d1145c9d489a8f82dc76d/cpp/include/cudf/round.hpp#L30-L36

Comment (bdice) [reply cpp/src/datetime/datetime_ops.cu:60]: On the other hand, keeping the name "round" emphasizes the alignment with "ceil" and "floor", all of which have corresponding function names in pandas and `std::chrono`.

File: python/cudf/cudf/core/series.py
Comment (bdice) [inline python/cudf/cudf/core/series.py:4690]: Formatting for readability.
```suggestion
        >>> dt_sr = cudf.Series([
        ...     "2001-01-01 00:04:45",
        ...     "2001-01-01 00:04:58",
        ...     "2001-01-01 00:05:04",
        ... ], dtype="datetime64[ns]")
```

File: python/cudf/cudf/core/indexed_frame.py
Comment (mayankanand007) [reply python/cudf/cudf/core/indexed_frame.py:706]: That's a good point @bdice! I will keep that in mind, thanks!

File: cpp/src/datetime/datetime_ops.cu
Comment (mayankanand007) [reply cpp/src/datetime/datetime_ops.cu:60]: I like the idea of renaming `rounding_kind` to `rounding_function` if that conveys a better message and if it solves the problem. Let me know!

File: python/cudf/cudf/core/index.py
Comment (vyasr) [inline python/cudf/cudf/core/index.py:1942]: Is there a reason to name this `field` when it's discussed as a frequency `freq` in the docstring?

Comment (vyasr) [reply python/cudf/cudf/core/index.py:1942]: FWIW [here's the pandas API for comparison](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DatetimeIndex.round.html).

File: python/cudf/cudf/core/index.py
Comment (bdice) [inline python/cudf/cudf/core/index.py:1952]: The pandas docs say:

> Index of the same type for a DatetimeIndex or TimedeltaIndex, or a Series with the same index for a Series.

In the pandas code, this is indicating that the return type is the same as the _input_ type. Is this function used in cuDF for `DatetimeIndex`, `TimedeltaIndex`, and `Series`? If this is just for `DatetimeIndex` then we can try something like a short description (suggested below). If multiple input types are possible, then I would fix these docstrings to align with the pandas docs (to indicate that the return type varies with the input type).

```suggestion
        DatetimeIndex
            Index containing rounded datetimes.
```

File: python/cudf/cudf/core/index.py
Comment (bdice) [inline python/cudf/cudf/core/index.py:1963]: It would be nice to have multiple output values that are not all identical. Maybe start at 11:59, with a frequency of days, and round to hours. Then all the 11:59s will round to 12:00 on different days? Or perhaps use an initial time of 11:59, frequency of 3 hours, and round to the nearest hour.

Also show the output of `>>> rng` before rounding, so that users can see the before/after more clearly.

You could adapt the examples from `Series.dt.round` if you want.

File: python/cudf/cudf/core/indexed_frame.py
Comment (bdice) [inline python/cudf/cudf/core/indexed_frame.py:719]: Deleting some extra spaces.
```suggestion
            Number of decimal places to round each column to. This parameter
            must be an int for a Series. For a DataFrame, a dict or a Series
            are also valid inputs. If an int is given, round each column to the
            same number of places. Otherwise dict and Series round to variable
            numbers of places. Column names should be in the keys if
            `decimals` is a dict-like, or in the index if `decimals` is a
            Series. Any columns not included in `decimals` will be left as is.
            Elements of `decimals` which are not columns of the input will be
            ignored.
```

File: python/cudf/cudf/core/indexed_frame.py
Comment (bdice) [inline python/cudf/cudf/core/indexed_frame.py:746]: Fixing syntax (missing wrapped prompt `... `).
```suggestion
        >>> df = cudf.DataFrame(
        ...     [(.21, .32), (.01, .67), (.66, .03), (.21, .18)],
        ...     columns=['dogs', 'cats'],
        ... )
```

File: python/cudf/cudf/core/indexed_frame.py
Comment (bdice) [inline python/cudf/cudf/core/indexed_frame.py:755]:
```suggestion
        of decimal places.
```

File: python/cudf/cudf/core/indexed_frame.py
Comment (bdice) [inline python/cudf/cudf/core/indexed_frame.py:766]:
```suggestion
        With a dict, the number of places for specific columns can be
        specified with the column names as keys and the number of decimal
        places as values.
```

File: python/cudf/cudf/core/indexed_frame.py
Comment (bdice) [inline python/cudf/cudf/core/indexed_frame.py:777]:
```suggestion
        Using a Series, the number of places for specific columns can be
        specified with the column names as the index and the number of
        decimal places as the values.
```

**Review by bdice — CHANGES_REQUESTED** [review]

File: python/cudf/cudf/core/index.py
Comment (mayankanand007) [reply python/cudf/cudf/core/index.py:1952]: This is a good point @bdice ! In this case, `pandas` uses the same documentation across `dt`, `DatetimeIndex`, and `TimedeltaIndex` for `ceil`, `floor` and `round`. This PR adds `round` for `DatetimeIndex` and a datetime `Series`.

Your suggestion is actually better here because we have specific docs for each method and this method is just for `DatetimeIndex` as you pointed out, I will make that change. Thanks!

**Review by bdice — APPROVED** [review]
The changeset looks good to me, thanks for iterating on this @mayankanand007! I think there are some opportunities to consolidate the public C++ API and Cython bindings. I will file an issue documenting my proposal when I get a chance.

I'd like to see less boilerplate code repeated between ceil/floor/round in the tests (both Python and C++). I am approving because the new tests for round are aligned with the existing ceil/floor tests. A refactor could be done later.

File: cpp/src/datetime/datetime_ops.cu
Comment (bdice) [reply cpp/src/datetime/datetime_ops.cu:60]: @jrhemstad Do you have a preference among these suggestions?

**Comment by mayankanand007** [comment]
rerun tests

File: cpp/tests/datetime/datetime_ops_test.cpp
Comment (vyasr) [inline cpp/tests/datetime/datetime_ops_test.cpp:938]: This could probably all go into a templated helper function that accepts the resolution and the associated cudf function. Same goes for ceil/floor, though, so I'm fine pushing that to a future refactoring PR.

File: cpp/tests/datetime/datetime_ops_test.cpp
Comment (vyasr) [inline cpp/tests/datetime/datetime_ops_test.cpp:930]:
```suggestion
  auto timestamps = host_val.first;
```

File: python/cudf/cudf/core/index.py
Comment (vyasr) [inline python/cudf/cudf/core/index.py:1874]: I'm in favor of changing this as I advocated for it in `round` below, but changing the parameter name here and for `floor` below (preexisting APIs) makes this PR a breaking change. @shwina do you see any issues? If not, we just need to change the tag.

Comment (vyasr) [reply python/cudf/cudf/core/index.py:1874]: We could also argue that this is a bug (inconsistent with pandas) rather than a break.

**Review by vyasr — COMMENTED** [review]
I have a couple of small comments, but otherwise this looks good! I'll hold off on final approval pending @shwina's thoughts on the parameter name change.

File: cpp/tests/datetime/datetime_ops_test.cpp
Comment (bdice) [reply cpp/tests/datetime/datetime_ops_test.cpp:938]: This is a good solution for the repeated boilerplate. I agree it would be better for a future PR.

File: python/cudf/cudf/core/index.py
Comment (bdice) [reply python/cudf/cudf/core/index.py:1874]: I’d say it’s a bug. Also with single-argument functions, keywords are much less common.

File: cpp/tests/datetime/datetime_ops_test.cpp
Comment (mayankanand007) [reply cpp/tests/datetime/datetime_ops_test.cpp:938]: Thanks for the reviews! I agree , with so much code in common, there is an opportunity to combine these implementations. Happy to take this up after this PR😄

File: python/cudf/cudf/core/index.py
Comment (shwina) [reply python/cudf/cudf/core/index.py:1874]: I think the parameter change is fine and desirable. Agree that this is a bug, not a breaking change.

**Review by vyasr — APPROVED** [review]
LGTM. @mayankanand007 you'll need to fix the style checks, but I assume that the tests will pass after.

**Comment by mayankanand007** [comment]
> LGTM. @mayankanand007 you'll need to fix the style checks, but I assume that the tests will pass after.

Thanks @vyasr ! the style failure is due to a known issue (#9870), waiting for that to get merged. Does this PR need more approvals?

File: cpp/src/datetime/datetime_ops.cu
Comment (jrhemstad) [reply cpp/src/datetime/datetime_ops.cu:60]: `rounding_function` sounds good

Comment (mayankanand007) [reply cpp/src/datetime/datetime_ops.cu:60]: @bdice does adding this, make this PR a breaking change? Do I need to change any labels?

Comment (bdice) [reply cpp/src/datetime/datetime_ops.cu:60]: Nope, this enum is in the `cudf::detail` namespace. Changes to private or detail APIs do not require a breaking label.

**Comment by mayankanand007** [comment]
@gpucibot merge

**Pull request merged**
